### ISSUE TITLE
migrate tool ignores parser

### ISSUE THREAD

**Issue by jdelStrother** [storybookjs/storybook#8091]
**Describe the bug**
When running `sb migrate` with a parser argument (eg `sb migrate --parser=flow storiesof-to-csf --glob 'components/stories/*.js'`), the parser is ignored.

**To Reproduce**
1. Create a stories file with flowtype annotations
2. Run `sb migrate --parser=flow storiesof-to-csf --glob 'components/stories/*.js'`

**Expected behavior**
The migrate tool converts my stories to CSF format

**Screenshots**
Failure due to babel parser not being able to parse flow types:
```
 ERR components/stories/midrollToolbar.js Transformation error
SyntaxError: Unexpected token, expected ";" (7:11)
    at _class.raise (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:776:15)
    at _class.unexpected (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:2079:16)
    at _class.semicolon (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:2063:40)
    at _class.parseVarStatement (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:4413:10)
    at _class.parseStatementContent (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:4012:21)
    at _class.parseStatement (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:3962:17)
    at _class.parseBlockOrModuleBlockBody (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:4513:23)
    at _class.parseBlockBody (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:4500:10)
    at _class.parseBlockBody (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:5462:44)
    at _class.parseTopLevel (/Users/jon/.config/yarn/global/node_modules/jscodeshift/node_modules/babylon/lib/index.js:3938:10)
```

**Additional context**

The parser option is passed here - https://github.com/storybookjs/storybook/blob/dfe9eb7efc73cd677f44c1ebb3f052ef68b4a3f9/lib/cli/bin/generate.js#L68 - and then ignored here - https://github.com/storybookjs/storybook/blob/dfe9eb7efc73cd677f44c1ebb3f052ef68b4a3f9/lib/cli/lib/migrate.js#L3.

I tried to naively fix it by passing the parserOption through to `runCodeMod`, but it dies at the prettier stage:

```
 ERR components/stories/midrollToolbar.js Transformation error (Cannot read property 'process' of undefined)
TypeError: Cannot read property 'process' of undefined
    at buffer (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:42536)
    at mGt (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:1429771)
    at e (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:393)
    at default (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:19587)
    at /Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:16
    at Object.prettierPlugins (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:139)
    at Module._compile (internal/modules/cjs/loader.js:816:30)
    at Module._compile (/Users/jon/Developer/web/webpack/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:827:10)
    at Object.newLoader [as .js] (/Users/jon/Developer/web/webpack/node_modules/pirates/lib/index.js:104:7)
```

**Comment by shilman** [comment]
@jdelStrother I'm a little confused. Is the option getting ignored, or getting recognized and then failing? It looks to me like it's getting passed properly in the code--I haven't set up a repro yet. Thanks!

**Comment by jdelStrother** [comment]
@shilman it's getting ignored.  cli/bin/generate.js does pass the migrate parameter through to cli/lib/migrate.js, but that then just ignores it https://github.com/storybookjs/storybook/blob/dfe9eb7efc73cd677f44c1ebb3f052ef68b4a3f9/lib/cli/lib/migrate.js#L3

**Comment by jdelStrother** [comment]
Something like this would fix the parser option getting ignored - 

```diff
import { listCodemods, runCodemod } from '@storybook/codemod';

- export async function migrate(migration, { configDir, glob, dryRun, list, rename, logger }) {
+ export async function migrate(migration, { configDir, glob, dryRun, list, rename, logger, parser }) {

  if (list) {
    listCodemods().forEach(key => logger.log(key));
  } else if (migration) {
-    await runCodemod(migration, { configDir, glob, dryRun, logger, rename });
+    await runCodemod(migration, { configDir, glob, dryRun, logger, rename, parser });
  } else {
    throw new Error('Migrate: please specify a migration name or --list');
  }
}
```

but that then fails in new and interesting ways inside the codemod - eg 
```
 ERR components/stories/midrollToolbar.js Transformation error (Cannot read property 'process' of undefined)
TypeError: Cannot read property 'process' of undefined
    at buffer (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:42536)
    at mGt (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:1429771)
    at e (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:393)
    at default (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:19587)
    at /Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:16
    at Object.prettierPlugins (/Users/jon/Developer/web/webpack/node_modules/prettier/parser-flow.js:1:139)
    at Module._compile (internal/modules/cjs/loader.js:816:30)
    at Module._compile (/Users/jon/Developer/web/webpack/node_modules/pirates/lib/index.js:99:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:827:10)
    at Object.newLoader [as .js] (/Users/jon/Developer/web/webpack/node_modules/pirates/lib/index.js:104:7)

```

**Comment by mbylstra** [comment]
There's another way you can run the `storiesof-to-csf` transform as described in https://www.npmjs.com/package/@storybook/codemod. 

Eg: `./node_modules/.bin/jscodeshift --parser=tsx -t ./node_modules/@storybook/codemod/dist/transforms/storiesof-to-csf.js . --ignore-pattern "node_modules|dist"`

This should bypass the problem of the `--parser=flow` argument not being passed to `jscodeshift`. There's still potentially further problems. I'm using this technique this to parse `xxxx.stories.tsx` files. The parser is working, but the transform itself appears to not work on the Typescript AST. It's removing the the `storiesOf` import statement, so it's doing something, but it's not doing anything more than that.

**Comment by shilman** [comment]
@mbylstra  somebody told me that it’s actually prettier that’s falling on TSX and not the codemod itself. And that if you comment out the prettier invocation in the package dist it will succeed. I’m on vacation this week so i haven’t had time to investigate. But if it’s easy for you to verify that, it would help give an extra data point for a proper fix

**Comment by mbylstra** [comment]
Yes, thanks - I also found that. But then I ran into further problems with the transform when it's run in tsx mode: I opened an issue about this: https://github.com/storybookjs/storybook/issues/9669

**Comment by shilman** [comment]
Fixed in #12453 and released in 6.0. Closing!
